### PR TITLE
Es6 proxy

### PR DESCRIPTION
**Pull request by dam0vm3nt**
Support for lists and maps.

### CODE REVIEW

File: lib/src/convert_es6_proxy.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+final JsObject _Unsupported = _polymerInteropDartES6['Unsupported'];
+
+/// Hooks for ES6 Proxies
+///
+/// Hooks for getting and setting properties and methods from and to dart objects.
+///
+JsFunction _initES6ListProxySupport() {
+  <String, Function>{
+    '_dartArrayGet': (List instance, index) =>
+        (index is num) ? convertToJs(instance[index]) : _Unsupported,
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: `num` => `int`

Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: Or actually just change `index` to have type `int`?

File: lib/src/convert_es6_proxy.dart [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    '_dartArrayPut': (List instance, int index, val) {
+      if (index is num) {
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: This is specified to be of type `int` above, which should be sufficient.

File: lib/polymer_interop_config.dart [hunk 3]
```diff
@@ -0,0 +1,27 @@
+library polymer_interop.polymer_interop_config;
```
Comment (jakemac53) [inline lib/polymer_interop_config.dart]: Lets stick this file under `lib/src`, and export it from `lib/polymer_interop.dart`.

File: lib/src/convert_es6_proxy.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        instance[index] = convertToDart(val);
+      } else {
+        return _Unsupported;
+      }
+    },
+    '_dartArrayLength': (List instance) => instance.length,
+    '_dartArraySplice': (List instance, int index, int howmany, List items) {
+      index = index == null ? 0 : index;
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: nit, `index ??= 0;`

File: lib/src/convert_es6_proxy.dart [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        (index is num) ? convertToJs(instance[index]) : _Unsupported,
+    '_dartArrayPut': (List instance, int index, val) {
+      if (index is num) {
+        instance[index] = convertToDart(val);
+      } else {
+        return _Unsupported;
+      }
+    },
+    '_dartArrayLength': (List instance) => instance.length,
+    '_dartArraySplice': (List instance, int index, int howmany, List items) {
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: Will want to add some good tests for this one as it is easy to get subtly wrong ;)

File: lib/src/convert_es6_proxy.dart [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      index = index == null ? 0 : index;
+      int end = (howmany == null) ? instance.length : index + howmany;
+
+      JsArray removed =
+          new JsArray.from(instance.sublist(index, end).map(convertToJs));
+
+      items = items == null ? [] : items;
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: nit, `items ??= [];`

File: lib/src/convert_es6_proxy.dart [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        return _Unsupported;
+      }
+    },
+    '_dartArrayLength': (List instance) => instance.length,
+    '_dartArraySplice': (List instance, int index, int howmany, List items) {
+      index = index == null ? 0 : index;
+      int end = (howmany == null) ? instance.length : index + howmany;
+
+      JsArray removed =
+          new JsArray.from(instance.sublist(index, end).map(convertToJs));
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: I think you want a call to `removeRange` here as well

File: lib/src/convert_es6_proxy.dart [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+      items = items == null ? [] : items;
+
+      instance.replaceRange(index, end, items.map(convertToDart));
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: I think you want `instance.insertAll(index, items.map(convertToDart));` here (combined with my other recommendation about using `removeRange`.

File: lib/src/convert_es6_proxy.dart [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+      return removed;
+    },
+    '_dartArraySlice': (List instance, [int begin, int end]) {
+      begin = begin == null ? 0 : begin;
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: same nit, `begin ??= 0;`

File: lib/src/convert_es6_proxy.dart [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      JsArray removed =
+          new JsArray.from(instance.sublist(index, end).map(convertToJs));
+
+      items = items == null ? [] : items;
+
+      instance.replaceRange(index, end, items.map(convertToDart));
+
+      return removed;
+    },
+    '_dartArraySlice': (List instance, [int begin, int end]) {
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.dart]: Not that slice (and maybe also splice) support negative numbers in JS but not in Dart. We need to handle that case, although it shouldn't be to difficult.

File: lib/src/convert_es6_proxy.html [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+  Polymer.PolymerInterop = Polymer.PolymerInterop || {};
+  Polymer.PolymerInterop.ES6 = Polymer.PolymerInterop.ES6 || {};
+
+
+
+  (function(ES6) {
+
+    ES6.Unsupported = {};
+
+    var isADartMethod = function() {
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.html]: maybe rename this, something like `_throwShouldBeOverridden`

File: lib/src/convert_es6_proxy.html [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+          var res = ES6._dartGetter(instance,propertyName);
+          if(res === ES6.Unsupported) {
+            return target[propertyName];
+          } else {
+            return res;
+          }
+        },
+        'set' : function(target,propertyName,value,proxy) {
+          if (ES6._dartSetter(instance,propertyName,value) === ES6.Unsupported) {
+            target[propertyName] = value;
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.html:78]: I don't think you actually want to assign a value here? I would throw an error instead.

File: lib/src/convert_es6_proxy.html [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+      return new Proxy({},{
+        'get' : function(target,propertyName,proxy) {
+
+          if (commonBehaviors.hasOwnProperty(propertyName)) {
+            return commonBehaviors[propertyName](instance);
+          }
+
+          var res = ES6._dartGetter(instance,propertyName);
+          if(res === ES6.Unsupported) {
+            return target[propertyName];
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.html:50]: This should probably throw, or possibly return `undefined`

File: lib/src/convert_es6_proxy.html [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    };
+
+    /**
+     * List ES6JsProxy support
+     */
+
+    var arrayCommonBehaviors = {
+      '__dartClass__' : function(instance) { return instance; },
+      '__isES6Proxy__' : function(instance) { return true; },
+      'splice' : function (instance) {
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.html]: There are a lot more things defined on `Array` other than these, but these are the most commonly used so I think its a decent start.

I don't know if we should try to implement the whole `Array` prototype or what, that would definitely be a pain to maintain as well....

I think we should make sure the failure case for trying to use one of the other methods be good at least.

File: lib/src/convert_es6_proxy.html [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    ES6.createES6JsProxyForArray = function(instance) {
+      return new Proxy([],{
+        'get' : function(target,propertyName,proxy) {
+
+          if (arrayCommonBehaviors.hasOwnProperty(propertyName)) {
+            return arrayCommonBehaviors[propertyName](instance);
+          }
+
+          if (isNaN(propertyName)) {
+            return target[propertyName];
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.html]: Same as above, I would throw or return undefined.

File: lib/src/convert_es6_proxy.html [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+          }
+
+          return ES6._dartArrayGet(instance,+propertyName);
+        },
+
+        'set' : function(target,propertyName,value,proxy) {
+          if (isNaN(propertyName)) {
+            target[propertyName] = value;
```
Comment (jakemac53) [inline lib/src/convert_es6_proxy.html]: Same as above, here I would throw

File: lib/src/convert_es6_proxy.dart (hunk 1 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.dart]: I still have to figure out this. An Array in JS-lang is sometime decorated with other properties that are not number. Dart interop for example adds some hiddent properties to store the dart equivalent object and Polymer too will if I remember.

But actually _dartArrayGet should be used only when index is an integer so it's right here to seti it as an int.

File: lib/polymer_interop_config.dart (hunk 3 above)
Comment (dam0vm3nt) [reply lib/polymer_interop_config.dart]: k

File: lib/src/convert_es6_proxy.dart (hunk 4 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.dart]: gotcha! this is nice. I've always hated trygraph for this kind of things.

File: lib/src/convert_es6_proxy.dart (hunk 8 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.dart]: Ok. remove and add should be more efficient.

File: lib/src/convert_es6_proxy.dart (hunk 10 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.dart]: Right, I forgot it. I'll add support for negative indexes too.

File: lib/src/convert_es6_proxy.html (hunk 11 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.html]: k

File: lib/src/convert_es6_proxy.html (hunk 12 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.html:78]: Sometimes an array get decorated with hidden properties. For example Dart Interop will add a cache for the corresponding Dart JsArray on the other side. I'm leaving this until I figure out all the use cases ( limiting to what it is need to work with polymer)

File: lib/src/convert_es6_proxy.html (hunk 13 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.html:50]: same as before. Also the get trap is used also to get the "constructor" and "**proto**" properties. For those properties I'll pass through to the underlining array.

File: lib/src/convert_es6_proxy.html (hunk 14 above)
Comment (dam0vm3nt) [reply lib/src/convert_es6_proxy.html]: Yes. Actually this work could eventually be ported inside dart interop in the end. Then it should support the whole Array / Map interface. But for now I want to implement only what it is needed for polymer to accept it as true js array.

**Comment by dam0vm3nt** [comment]
superseded by https://github.com/dart-lang/polymer_interop/pull/47

**Pull request closed without merging**